src/datatypes/datetime.coffee

```diff
@@ -676,6 +570,243 @@ DateTime.prototype.isSamePrecision = Date.prototype.isSamePrecision = (other) ->
+    if @[field]? and other[field]?
+      # if this value is less than the other return with true. this is before other
+      if @[field] < other[field]
+        return true
+      # if this value is greater than the other return with false. this is after
+      else if @[field] > other[field]
+        return false
+      # execution continues if the values are the same
+
+    # if both dont have this precision, return true if precision is not defined
+    else if !@[field]? and !other[field]?
+      if !precision?
+        return true
+      else # we havent met precision yet
+        return null
+
+    # otherwise they have inconclusive precision, return null
+    else
+      return null
+
+    # if precision is defined and we have reached expected precision, we can leave the loop
+    break if precision? and precision is field
+
+  # if we made it here, then all fields matched and they are same
+  true
+
+DateTime.prototype.sameOrAfter = Date.prototype.sameOrAfter = (other, precision) ->
+  if not((other.isDate) or (other.isDateTime))
+    return null
+  else if @.isDate and other.isDateTime
+    return @getDateTime().sameOrAfter(other, precision)
+  else if @.isDateTime and other.isDate
+    other = other.getDateTime()
+
+  if precision? && @constructor.FIELDS.indexOf(precision) < 0
+    throw new Error("Invalid precision: #{precision}")
+
+  # make a copy of other in the correct timezone offset if they don't match.
+  if (@timezoneOffset != other.timezoneOffset)
+    other = other.convertToTimezoneOffset(@timezoneOffset)
+
+  for field in @constructor.FIELDS
+    # if both have this precision defined
+    if @[field]? and other[field]?
+      # if this value is greater than the other return with true. this is after other
+      if @[field] > other[field]
+        return true
+      # if this value is greater than the other return with false. this is before
+      else if @[field] < other[field]
+        return false
+      # execution continues if the values are the same
+
+    # if both dont have this precision, return true if precision is not defined
+    else if !@[field]? and !other[field]?
+      if !precision?
+        return true
+      else # we havent met precision yet
+        return null
+
+    # otherwise they have inconclusive precision, return null
+    else
+      return null
+
+    # if precision is defined and we have reached expected precision, we can leave the loop
+    break if precision? and precision is field
+
+  # if we made it here, then all fields matched and they are same
+  true
+
+DateTime.prototype.before = Date.prototype.before = (other, precision) ->
+  if not((other.isDate) or (other.isDateTime))
+    return null
+  else if @.isDate and other.isDateTime
+    return @getDateTime().before(other, precision)
+  else if @.isDateTime and other.isDate
+    other = other.getDateTime()
+
+  if precision? && @constructor.FIELDS.indexOf(precision) < 0
+    throw new Error("Invalid precision: #{precision}")
```

Comment:
We could also enumerate the valid precisions in this error message, which might be helpful.

Reply:
The CQL to ELM translator rejects invalid precisions. The only case that we will see this is if someone uploads a CQL 1.2 measure that uses `week` precision. Now that I think about it, I will create a test to make sure the right list is being used to reject the time level precisions on Date objects.